src/Language/DifferentialDatalog/D3log.hs

```diff
@@ -0,0 +1,124 @@
+--
+-- For each relation 'R[T]' that appears in the head of at least one rule
+-- with location annotation, we introduce a fresh output relation
+-- '__out_R[(Option<D3logLocationId>, T)]' and replaces all occurrences
+-- 'R' in the head of a rule with '__out_R'.
+--
+-- Returns modified program along with a map from newly introduced output
+-- relation names to corresponding input relations: '__out_R -> R'
+processD3logAnnotations :: (?cfg::Config) => DatalogProgram -> (DatalogProgram, M.Map String String)
+processD3logAnnotations d | confD3log ?cfg == False = (d, M.empty)
+                          | otherwise = 
+    (d { progRelations = rels', progRules = rules' }, M.fromList $ map (\n -> (outRelName n, n)) $ S.toList loc_rels)
+    where
+    -- Find annotated relations.
+    loc_rels = progLocalizedRelations d
+    
+    -- For each localized relation 'R[T]', convert 'R'
+    -- into an input relation and generate a new output relation
+    -- 'output relation __out_R[(Option<D3logLocationId>, T)]'.
+    rels' = foldl' (\rels rname ->
+                     let rel = progRelations d M.! rname
+                         rel_in = rel{relRole = RelInput}
+                         rel_out = rel{
+                             relName = outRelName $ relName rel,
+                             relRole = RelOutput,
+                             relType = tTuple [locationSpecifierType, relType rel]
+                         }
+                     in M.insert (name rel_in) rel_in $
+                        M.insert (name rel_out) rel_out $
+                        M.delete rname rels)
+                   (progRelations d) loc_rels
+
+    -- Rewrite rules:
+    --  'R[x] :- Body.' --> '__out_R[(None, x)] :- Body.'
+    --  and
+    --  'R[x] @n :- Body.' --> '__out_R[(Some{n}, x)] :- Body.'
+    rules' = map (\rule ->
+                   let rlhead = head $ ruleLHS rule
+                       rname = atomRelation $ lhsAtom rlhead
+                       loc_expr = case lhsLocation rlhead of
+                                       Nothing -> eNone $ tUser lOCATION_TYPE []
+                                       Just l  -> eSome l $ tUser lOCATION_TYPE []
+                       val' = eTuple [loc_expr, atomVal $ lhsAtom rlhead]
+                       atom' = (lhsAtom rlhead) {
+                           atomRelation = outRelName rname,
+                           atomVal = val'
+                       }
+                       rlhead' = rlhead {
+                           lhsAtom     = atom',
+                           lhsLocation = Nothing
+                       } in
+                   if S.member rname loc_rels
+                   then rule{ ruleLHS = [rlhead'] }
+                   else rule)
+             $ progRules d
+
+-- 'Option<D3logLocationId>'
+locationSpecifierType :: Type
+locationSpecifierType = tUser oPTION_TYPE [tUser lOCATION_TYPE []]
+
+outRelName :: String -> String
+outRelName rel = scoped rel_scope $ "__out_" ++ rel_name
```

Comment:
I would be more comfortable if we had a fresh name allocator

Reply:
Yes, this is a recurring problem in the compiler. The reason this is not completely trivial is because we'd have to thread state through otherwise purely function code, but it's probably still worth doing. Do you want to submit an issue?